cmd_migrate_db.go

```diff
@@ -0,0 +1,705 @@
+	"github.com/jessevdk/go-flags"
+	"github.com/lightninglabs/lndinit/migratekvdb"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/kvdb"
+	"github.com/lightningnetwork/lnd/kvdb/postgres"
+	"github.com/lightningnetwork/lnd/kvdb/sqlbase"
+	"github.com/lightningnetwork/lnd/kvdb/sqlite"
+	"github.com/lightningnetwork/lnd/lncfg"
+	"github.com/lightningnetwork/lnd/watchtower/wtdb"
+)
+
+var (
+	// alreadyMigratedKey is the key under which we add a tag in the target/
+	// destination DB after we've successfully and completely migrated it
+	// from a source DB.
+	alreadyMigratedKey = []byte("data-migration-already-migrated")
+
+	// defaultDataDir is the default data directory for lnd.
+	defaultDataDir = filepath.Join(btcutil.AppDataDir("lnd", false), "data")
+)
+
+const (
+	// walletMetaBucket is the name of the meta bucket in the wallet db
+	// for the wallet ready marker.
+	walletMetaBucket = "lnwallet"
+
+	// walletReadyKey is the key in the wallet meta bucket for the wallet
+	// ready marker.
+	walletReadyKey = "ready"
+)
+
+// Bolt is the configuration for a bolt database.
+type Bolt struct {
+	DBTimeout time.Duration `long:"dbtimeout" description:"Specify the timeout value used when opening the database."`
+	DataDir   string        `long:"data-dir" description:"Lnd data dir where bolt dbs are located."`
+	TowerDir  string        `long:"tower-dir" description:"Lnd watchtower dir where bolt dbs for the watchtower server are located."`
+}
+
+// Sqlite is the configuration for a sqlite database.
+type Sqlite struct {
+	DataDir  string         `long:"data-dir" description:"Lnd data dir where sqlite dbs are located."`
+	TowerDir string         `long:"tower-dir" description:"Lnd watchtower dir where sqlite dbs for the watchtower server are located."`
+	Config   *sqlite.Config `group:"sqlite-config" namespace:"sqlite-config" description:"Sqlite config."`
+}
+
+// SourceDB represents the source database, which can only be bolt for now.
+type SourceDB struct {
+	Backend string `long:"backend" description:"The source database backend." choice:"bolt"`
+	Bolt    *Bolt  `group:"bolt" namespace:"bolt" description:"Bolt settings."`
+}
+
+// DestDB represents the destination database, which can be either postgres or
+// sqlite.
+type DestDB struct {
+	Backend  string           `long:"backend" description:"The destination database backend." choice:"postgres" choice:"sqlite"`
+	Postgres *postgres.Config `group:"postgres" namespace:"postgres" description:"Postgres settings."`
+	Sqlite   *Sqlite          `group:"sqlite" namespace:"sqlite" description:"Sqlite settings."`
+}
+
+// Init should be called upon start to pre-initialize database for sql
+// backends. If max connections are not set, the amount of connections will be
+// unlimited however we only use one connection during the migration.
+func (db *DestDB) Init() error {
+	switch {
+	case db.Backend == lncfg.PostgresBackend:
+		sqlbase.Init(db.Postgres.MaxConnections)
```

Comment:
In `lnd`, we also apply some extra config settings to speed things up (like activating the WAL): https://github.com/lightningnetwork/lnd/blob/6a3845b79da835071bf03085007b4a6e76c009d3/kvdb/sqlite/db.go#L33-L59.

This is also a chance to activate the incremental vacuum option, as you can't activate it _after_ a db has tables already. We should also add this to that lnd section I linked above: https://www.sqlite.org/pragma.html#pragma_auto_vacuum

Reply:
Nvm, I realize that we'll use `kvdb.Open` which will go to `walletdb.Open` to ensure that all the proper configs are set.